apps/anoma_node/lib/node/transaction/ordering.ex

```diff
@@ -444,4 +1150,358 @@ defmodule Anoma.Node.Transaction.Ordering do
+          GenServer.from(),
+          t()
+        ) :: any()
+  defp handle_shard_operation(type, tx_id, key, value, from, state) do
+    case Map.fetch(state.tx_id_to_height, tx_id) do
+      {:ok, height} ->
+        case get_shard_pid(key, state) do
+          {:ok, shard_pid} ->
+            # Have height and shard pid, check reservation
+            case get_reservation_for_key(tx_id, key, type, state) do
+              {:ok, _} ->
+                # Have reservation, proceed with operation
+                Task.start(fn ->
+                  result =
+                    case type do
+                      :read -> Shard.read(shard_pid, key, height)
+                      :write -> Shard.write(shard_pid, key, value, height)
+                    end
+
+                  GenServer.reply(from, result)
+                end)
+
+              :error ->
+                # No reservation found, queue request to wait for acquisition
+                block_spawn(
+                  tx_id,
+                  fn ->
+                    block_for_reservation(
+                      type,
+                      tx_id,
+                      key,
+                      value,
+                      from,
+                      shard_pid,
+                      height,
+                      state.node_id
+                    )
+                  end,
+                  state.node_id
+                )
+            end
+
+          {:error, reason} ->
+            Logger.error(
+              "Failed to get shard PID for key #{inspect(key)} in handle_shard_operation: #{inspect(reason)}"
+            )
+
+            # Block until height is known (will implicitly retry shard PID lookup upon retry)
+            block_spawn(
+              tx_id,
+              fn ->
+                blocking_shard_operation(
+                  type,
+                  state.node_id,
+                  tx_id,
+                  key,
+                  value,
+                  from
+                )
+              end,
+              state.node_id
+            )
+        end
+
+      :error ->
+        # Block until height is known
+        block_spawn(
+          tx_id,
+          fn ->
+            blocking_shard_operation(
+              type,
+              state.node_id,
+              tx_id,
+              key,
+              value,
+              from
+            )
+          end,
+          state.node_id
+        )
+    end
+  end
+
+  # Helper function to wait for reservations to be acquired before proceeding with a read/write.
+  @spec block_for_reservation(
+          :read | :write,
+          binary(),
+          binary(),
+          any() | nil,
+          GenServer.from(),
+          pid(),
+          integer(),
+          String.t()
+        ) :: :ok
+  defp block_for_reservation(
+         type,
+         tx_id,
+         key,
+         # nil for reads
+         value,
+         from,
+         shard_pid,
+         height,
+         node_id
+       ) do
```

Comment:
This seems like a lot of arguments, maybe we can cut it down by some structs perhaps? Can those natrually occur?